### PR TITLE
Ledger signer 221

### PR DESCRIPTION
**Pull request by PhilippeR26**
## Motivation and Resolution
Ledger has released a new version 2.2.1 of the Starknet APP.
It includes improvements, including a comprehensive display of the details of:
- V1 & V3 transactions.
- V1& V3 deployments.

Starknet.js is now able to use APP v1.1.1 & v2.2.1.

- For users still using v1.1.1 :
No change of the code :
```typescript
const myLedgerSigner = new LedgerSigner(myLedgerTransport, 0);
```
`LedgerSigner` and `LedgerSigner111` classes are equivalent.
- For new users, using v2.2.1 to create a new account :
```typescript
const myLedgerSigner = new LedgerSigner221(myLedgerTransport, 0);
```
They can also use the v1.1.1 signer ; they will only not have access to the display of transaction details.

- For new users, using v1.1.1 to create a new account that will be natively compatible with v2.2.1 :
```typescript
const myLedgerSigner = new LedgerSigner111(myLedgerTransport, 0, undefined, getLedgerPathBuffer221);
```
- For users that have created an account with V1.1.1, and are now using v2.2.1 :
```typescript
const myLedgerSigner = new LedgerSigner221(myLedgerTransport, 0, getLedgerPathBuffer111);
```

## Development related changes
`LedgerSigner` and `getLedgerPathBuffer` are still there, and are related to v1.1.1 APP.  
They are now also named `LedgerSigner111` and `getLedgerPathBuffer111`.

`LedgerSigner221` and `getLedgerPathBuffer221` have been created to handle v2.2.1.

As several functions are common for both signers, the `LedgerSigner221` is an extension of the `LedgerSigner111`.

Some maths included in `hashFeeField` (utils/hash/transactionHash/v3.ts) are necessary, and have been isolated in exportable functions.

The new signer has been tested successfully in Node ( https://github.com/PhilippeR26/starknet.js-workshop-typescript/blob/main/src/scripts/ledgerNano/6.testLedgerAccount221.ts ) and in a demo DAPP ( https://github.com/PhilippeR26/Starknet-Ledger-Wallet )

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

File: src/signer/ledgerSigner211.ts [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+   * // res = {hash:
+   * //  signature:
+   * }
+   * ```
+   */
+  async signTxV1(
+    txDetails: V2InvocationsSignerDetails,
+    calls: Call[]
+  ): Promise<{ hash: bigint; signature: Signature }> {
+    // APDU 0 for path
```
Comment (ivpavici) [inline src/signer/ledgerSigner211.ts:237]: what is APDU ? :)

File: __tests__/utils/ethSigner.test.ts [hunk 2]
```diff
@@ -354,12 +355,25 @@ describe('Ethereum signer', () => {
 describe('Ledger Signer', () => {
   // signature of Ledger can't be tested automatically.
```
Comment (tabaktoni) [inline __tests__/utils/ethSigner.test.ts:356]: Maybe there is some way with virtual ledger like
https://developers.ledger.com/docs/ledger-live/discover/integration/wallet-api/simulator/introduction

File: src/constants.ts [hunk 3]
```diff
@@ -82,3 +82,7 @@ export const SNIP9_V1_INTERFACE_ID =
   '0x68cfd18b92d1907b8ba3cc324900277f5a3622099431ea85dd8089255e4181';
 export const SNIP9_V2_INTERFACE_ID =
   '0x1d1144bb2138366ff28d8e9ab57456b1d332ac42196230c3a602003c89872';
+
+// Ledger signer
+export const HARDENING_BYTE = Number('0x80');
```
Comment (tabaktoni) [inline src/constants.ts:87]: nitpick, Can we store this as a static constant
```suggestion
// 0x80
export const HARDENING_BYTE = 128;
```

File: src/constants.ts [hunk 4]
```diff
@@ -82,3 +82,7 @@ export const SNIP9_V1_INTERFACE_ID =
...
+export const HARDENING_4BYTES = BigInt('0x80000000');
```
Comment (tabaktoni) [inline src/constants.ts:88]: nitpick , Same
```suggestion
// 0x80000000
export const HARDENING_4BYTES = 2147483648n
```

File: src/utils/hash/transactionHash/v3.ts [hunk 5]
```diff
@@ -24,17 +24,33 @@ export function hashDAMode(nonceDAMode: BigNumberish, feeDAMode: BigNumberish) {
   return (BigInt(nonceDAMode) << DATA_AVAILABILITY_MODE_BITS) + BigInt(feeDAMode);
 }
 
-export function hashFeeField(tip: BigNumberish, bounds: ResourceBounds) {
-  const L1Bound =
+export function encodeResourceBoundsL1(bounds: ResourceBounds): bigint {
```
Comment (tabaktoni) [inline src/utils/hash/transactionHash/v3.ts:27]: maybe add desc, and note what this encoding is for

**Review by tabaktoni — APPROVED** [review]
There are a couple of nitpicks but in general LGTM!

We would need to refactor this soon when v1 tx gets deprecated with RPC 0.8

**Comment by PhilippeR26** [comment]
Ledger informed me that they are preparing a new version, including a human readable account deployment .
So, hold on for this PR. I will update it once the new Ledger APP available, including at the same time your comments.

File: __tests__/utils/ethSigner.test.ts (hunk 2 above)
Comment (PhilippeR26) [reply __tests__/utils/ethSigner.test.ts:356]: Seems not to be very reliable. The dev (a French guy) that coded the APP has no NanoX, tested it with success Speculos simulator.
On my side, I found several problems with my own NanoX. So, this simulator is useful for inital tests, but is not very reliable...

File: src/constants.ts (hunk 3 above)
Comment (PhilippeR26) [reply src/constants.ts:87]: Was just to highlight that it's just the higher bit.

Comment (PhilippeR26) [reply src/constants.ts:87]: Modified.

File: src/constants.ts (hunk 4 above)
Comment (PhilippeR26) [reply src/constants.ts:88]: Modified

File: src/utils/hash/transactionHash/v3.ts (hunk 5 above)
Comment (PhilippeR26) [reply src/utils/hash/transactionHash/v3.ts:27]: Added.

**Comment by PhilippeR26** [comment]
The final version of the Ledger APP is 2.2.1.
The code has been updated in accordance.
Reviewers requests are implemented.
Ready for final review.

File: www/docs/guides/signature.md [hunk 6]
```diff
@@ -228,15 +228,15 @@ import type Transport from '@ledgerhq/hw-transport'; // type for the transporter
 In a Web DAPP, take care that some browsers are not compatible (FireFox, ...), and that the Bluetooth is not working in all cases and in all operating systems.
 
 :::note
-The last version of the Ledger Starknet APP (v1.1.1) only supports blind signing of the hash of your action. Sign only hashes from a code that you trust.
+The last version of the Ledger Starknet APP (v2.2.1) supports explained V1 (ETH) & V3 (STRK) transactions & deploy accounts. For a class declaration or a message, you will have to blind sign a hash ; sign only hashes from a code that you trust. Do not forget to Enable `Bind signing` in the APP settings.
```
Comment (penovicp) [inline www/docs/guides/signature.md:231]:
```suggestion
The last version of the Ledger Starknet APP (v2.2.1) supports explained V1 (ETH) & V3 (STRK) transactions & deploy accounts. For a class declaration or a message, you will have to blind sign a hash ; sign only hashes from a code that you trust. Do not forget to Enable `Blind signing` in the APP settings.
```
Typo?

**Review by penovicp — APPROVED** [review]
One possible typo question, otherwise lgtm

File: www/docs/guides/signature.md [hunk 7]
```diff
@@ -260,3 +260,18 @@ const resp = await myLedgerTransport.send(Number('0x5a'), 0, 0, 0);
 const appVersion = resp[0] + '.' + resp[1] + '.' + resp[2];
 console.log('version=', appVersion);
 ```
+
+:::note
+You have also in Starknet.js a signer for the old v1.1.1 Ledger Starknet APP.
```
Comment (ivpavici) [inline www/docs/guides/signature.md:265]:
```suggestion
You also have in Starknet.js a signer for the old v1.1.1 Ledger Starknet APP.
```

File: src/signer/ledgerSigner221.ts [hunk 8]
```diff
@@ -0,0 +1,671 @@
...
+   * // result = Signature { r: 611475243393396148729326917410546146405234155928298353899191529090923298688n,
+   * // s: 798839819213540985856952481651392652149797817551686626114697493101433761982n,
+   * // recovery: 0}
+   * ```
+   */
+  public async signDeployAccountTransaction(
+    details: DeployAccountSignerDetails
+  ): Promise<Signature> {
+    const compiledConstructorCalldata = CallData.compile(details.constructorCalldata);
+    /*     const version = BigInt(details.version).toString(); */
```
Comment (ivpavici) [inline src/signer/ledgerSigner221.ts:173]: leftover?

**Review by ivpavici — COMMENTED** [review]
minor suggestions, not very important

**Comment by PhilippeR26** [comment]
@ivpavici All corrected and committed.

**Review by ivpavici — APPROVED** [review]
lgtm!

**Pull request merged**